lib/resourcebuilder/podspec.go

```diff
@@ -45,3 +45,78 @@ func updatePodSpecWithProxy(podSpec *corev1.PodSpec, containerNames []string, ht
 	return nil
 
 }
+
+// updatePodSpecWithInternalLoadBalancerKubeService mutates the input podspec by setting the KUBERNETES_SERVICE_HOST to the internal
+// loadbalancer endpoint and the KUBERNETES_SERVICE_PORT to the specified port
+func updatePodSpecWithInternalLoadBalancerKubeService(podSpec *corev1.PodSpec, containerNames []string, internalLoadBalancerHost, internalLoadBalancerPort string) error {
+	hasInternalLoadBalancer := len(internalLoadBalancerHost) > 0
+	if !hasInternalLoadBalancer {
+		return nil
+	}
+
+	for _, containerName := range containerNames {
+		found := false
+		for i := range podSpec.Containers {
+			if podSpec.Containers[i].Name != containerName {
+				continue
+			}
+			found = true
+
+			podSpec.Containers[i].Env = setKubeServiceValue(podSpec.Containers[i].Env, internalLoadBalancerHost, internalLoadBalancerPort)
+		}
+		for i := range podSpec.InitContainers {
+			if podSpec.InitContainers[i].Name != containerName {
+				continue
+			}
+			found = true
+
+			podSpec.InitContainers[i].Env = setKubeServiceValue(podSpec.Containers[i].Env, internalLoadBalancerHost, internalLoadBalancerPort)
+		}
+
+		if !found {
+			return fmt.Errorf("requested injection for non-existent container: %q", containerName)
+		}
+	}
+
+	return nil
+}
+
+// setKubeServiceValue replaces values if they are present and adds them if they are not
+func setKubeServiceValue(in []corev1.EnvVar, internalLoadBalancerHost, internalLoadBalancerPort string) []corev1.EnvVar {
+	ret := []corev1.EnvVar{}
+
+	portVal := "443"
```

Comment:
We should switch on scheme, so we default to 80 for the crazy folks that configure an HTTP LB.  Which is [techincally allowed][1] today.

[1]: https://github.com/openshift/api/blob/0ba2c3658da6cade27a5066575cd8446eb03914c/config/v1/types_infrastructure.go#L74

Reply:
or ensure that `KUBERNETES_SERVICE_PORT` is unset when `internalLoadBalancerPort == ""`, and leave the defaulting to the library code, which hopefully does not require an explicit `KUBERNETES_SERVICE_PORT`.  But I can check...

Reply:
[The docs][1] suggest that separate variables exist for `KUBERNETES_SERVICE_PORT` and `KUBERNETES_SERVICE_PORT_HTTPS`.  Might be recent, because I don't see any `KUBERNETES_SERVICE_PORT_HTTPS` matches in the CVO codebase.

[1]: https://kubernetes.io/docs/concepts/services-networking/connect-applications-service/#environment-variables

Reply:
Dunno what those docs are about, maybe they're aspirational.  But current CVO has:

```console
$ git --no-pager grep 'KUBERNETES_SERVICE_PORT must' origin/master
origin/master:vendor/k8s.io/client-go/rest/config.go:var ErrNotInCluster = errors.New("unable to load in-cluster configuration, KUBERNETES_SERVICE_HOST and KUBERNETES_SERVICE_PORT must be defined")
```

So back to my "we should switch the default depending on the configured scheme" suggestion.

Reply:
it is not allowed. The kube-apiserver cannot run in http. Those people will die out.

Reply:
We're on 6443, right?

Reply:
most folks will explicitly set `6443`, yes.  An `https://{host}` URI without an explicit port means "443" though, so this default makes sense to me there.

Reply:
openshift/api#704 now forbids `http`.  Once that's approved I will be satisfied with the hard-coded 443 here.

Reply:
[in the pipe][1].  Marking this thread resolved.

[1]: https://github.com/openshift/api/pull/704#issuecomment-662678578